port/zos390/omrvmem.c

```diff
@@ -1296,29 +1300,33 @@ omrvmem_default_large_page_size_ex(struct OMRPortLibrary *portLibrary, uintptr_t
 	}
 #endif /* OMR_ENV_DATA64 */
 
-	while (0 != PPG_vmem_pageSize[index]) {
-		if (0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode)) {
-			if ((ONE_M == PPG_vmem_pageSize[index]) &&
-				(0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE & PPG_vmem_pageFlags[index]))
-			) {
-				break;
+	for (int i = 0; ((0 != PPG_vmem_pageSize[i]) && (ONE_M >= PPG_vmem_pageSize[i])); i++) {
+		if (ONE_M == PPG_vmem_pageSize[i]) {
+			if (0 != (OMRPORT_VMEM_PAGE_FLAG_PAGEABLE & PPG_vmem_pageFlags[i])) {
+				indexPageable = i;
 			}
 #if defined(OMR_ENV_DATA64)
-		} else {
-			if ((ONE_M == PPG_vmem_pageSize[index]) &&
-				(0 != (OMRPORT_VMEM_PAGE_FLAG_FIXED & PPG_vmem_pageFlags[index]))
-			) {
-				break;
+			else if (0 != (OMRPORT_VMEM_PAGE_FLAG_FIXED & PPG_vmem_pageFlags[i])) {
+				index = i;
 			}
-#endif
+#endif /* OMR_ENV_DATA64 */
 		}
-		index++;
 	}
+
+	/* 
+	 * When mode is OMRPORT_VMEM_MEMORY_EXECUTE we will only consider pagable large pages.
+	 * When flag is OMRPORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE, we prefer pageable 1M large pages unless they're not provisioned.
+	 */
+	if ((0 != (OMRPORT_VMEM_MEMORY_MODE_EXECUTE & mode)) ||
+	   (TRUE == pageablePreferableRequested && (-1 != indexPageable))) {
+		index = indexPageable;
+	}
+
```

Comment:
if you add check for "not found" here you would not need to check for -1 inside:
```
    if (-1 == index) {
        if (NULL != pageSize) {
            *pageSize = PPG_vmem_pageSize[index];
        }
        if (NULL != pageFlags) {
            *pageFlags = PPG_vmem_pageFlags[index];
        }
    }
```

Reply:
sorry for typo comment above, fixed now